libs/langchain/langchain/schema/runnable/history.py

```diff
@@ -0,0 +1,254 @@
+from __future__ import annotations
+
+import asyncio
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Dict,
+    List,
+    Optional,
+    Sequence,
+    Type,
+    Union,
+)
+
+from langchain.load import load
+from langchain.pydantic_v1 import BaseModel, create_model
+from langchain.schema.chat_history import BaseChatMessageHistory
+from langchain.schema.runnable.base import Runnable, RunnableBindingBase, RunnableLambda
+from langchain.schema.runnable.config import RunnableConfig
+from langchain.schema.runnable.passthrough import RunnablePassthrough
+from langchain.schema.runnable.utils import (
+    ConfigurableFieldSpec,
+    get_unique_config_specs,
+)
+
+if TYPE_CHECKING:
+    from langchain.callbacks.tracers.schemas import Run
+    from langchain.schema.messages import BaseMessage
+
+MessagesOrDictWithMessages = Union[Sequence["BaseMessage"], Dict[str, Any]]
+
+
+class RunnableWithMessageHistory(RunnableBindingBase):
+    """
+    A runnable that manages chat message history for another runnable.
+
+    Base runnable must have inputs and outputs that can be converted to a list of
+        BaseMessages.
+
+    RunnableWithMessageHistory must always be called with a config that contains
+        {"configurable": {"thread_id": "<THREAD_ID>"}} and can optionally include
+        {"configurable": {"thread_id": "<THREAD_ID>", "user_id": "<USER_ID>"}}.
+
+    Example:
+        .. code-block:: python
+
+            from langchain.schema.prompts import ChatPromptTemplate, MessagesPlaceholder
+            from langchain.chat_models import ChatAnthropic
+            from langchain.schema.runnables import RunnableWithMessageHistory
+
+    """
+
+    factory: Callable[[str], BaseChatMessageHistory]
+    input_messages_key: Optional[str] = None
+    output_messages_key: Optional[str] = None
+    message_history_key: Optional[str] = None
+
+    def __init__(
+        self,
+        runnable: Runnable[
+            MessagesOrDictWithMessages,
+            Union[str, BaseMessage, MessagesOrDictWithMessages],
+        ],
+        factory: Callable[[str], BaseChatMessageHistory],
```

Comment:
What do you think about changing the factory signature to:

```factory: Callable[[dict], BaseChatMessageHistory]```

It might be annoying client side to parse the string?

Reply:
Replied in other thread